src/charm.py

```diff
@@ -8,18 +8,136 @@
+            f"http://kratos.{self.model.name}.svc.cluster.local:4433/sessions/whoami"
+        )
+
+        self.service_patcher = KubernetesServicePatch(
+            self, [("oathkeeper-api", OATHKEEPER_API_PORT)]
+        )
+
+        self.framework.observe(self.on.oathkeeper_pebble_ready, self._on_oathkeeper_pebble_ready)
+
+    @property
+    def _oathkeeper_layer(self) -> Layer:
+        """Returns a pre-configured Pebble layer."""
+        layer_config = {
+            "summary": "oathkeeper-operator layer",
+            "description": "pebble config layer for oathkeeper-operator",
+            "services": {
+                self._container_name: {
+                    "override": "replace",
+                    "summary": "Oathkeeper Operator layer",
+                    "command": f"oathkeeper serve -c {self._oathkeeper_config_path}",
+                    "startup": "enabled",
+                }
+            },
+            "checks": {
+                "alive": {
+                    "override": "replace",
+                    "http": {"url": f"http://localhost:{OATHKEEPER_API_PORT}/health/alive"},
+                },
+                "ready": {
+                    "override": "replace",
+                    "http": {"url": f"http://localhost:{OATHKEEPER_API_PORT}/health/ready"},
+                },
+            },
+        }
+        return Layer(layer_config)
+
+    @property
+    def _oathkeeper_service_is_created(self) -> bool:
+        try:
+            self._container.get_service(self._container_name)
+        except (ModelError, RuntimeError):
+            return False
+        return True
+
+    def _render_access_rules_file(self) -> str:
+        """Render the access rules file."""
+        with open("templates/access-rules.yaml.j2", "r") as file:
+            template = Template(file.read())
+
+        rendered = template.render(
+            kratos_login_url=self._kratos_login_url,
+            return_to="http://default-url.com",
+        )
+        return rendered
+
+    def _render_conf_file(self) -> str:
+        """Render the Oathkeeper configuration file."""
+        with open("templates/oathkeeper.yaml.j2", "r") as file:
+            template = Template(file.read())
+
+        rendered = template.render(
+            kratos_session_url=self._kratos_session_url,
+            kratos_login_url=self._kratos_login_url,
+        )
+        return rendered
+
+    def _on_oathkeeper_pebble_ready(self, event: PebbleReadyEvent) -> None:
+        """Event Handler for pebble ready event."""
+        if not self._container.can_connect():
+            event.defer()
+            logger.info("Cannot connect to Oathkeeper container. Deferring the event.")
+            self.unit.status = WaitingStatus("Waiting to connect to Oathkeeper container")
+            return
+
+        self.unit.status = MaintenanceStatus("Configuring the container")
+
+        self._container.add_layer(self._container_name, self._oathkeeper_layer, combine=True)
```

Comment:
dummy question, [here](https://github.com/canonical/oathkeeper-operator/pull/7/files#diff-b9ed39bbc9c0387bd3e07da31d13373745534a1cd723d3e292c73496b12e307cR118) we call \event.defer()\ and then we return 
i expect the code to flow back to this function and re add the layer, what does that imply?aka can we keep adding the very same layer?or we do it because the `self._container` gets reset due to the charm  being rerun from scratch?

(forgive me, forward me to the docs without any problem, just need some extra context really)